common-content/en/module/decomposition/why-types/index.md

```diff
@@ -0,0 +1,139 @@
+
+This gets even harder when code is only sometimes executed. For instance, consider this NodeJS program:
+
+```js
+import process from "node:process";
+import readline from "node:readline";
+
+const rl = readline.createInterface({
+  input: process.stdin,
+  output: process.stdout,
+});
+
+rl.question("What URL should we fetch?\n> ", async (url) => {
+  const response = await fetch(url);
+  if (!response.ok) {
+    if (response.body.toLowerCase().includes("permission")) {
+      console.error("You didn't have permission to get that URL");
+    } else {
+      console.error(`The request failed - body: ${response.body}`);
+    }   
+    process.exit(1);
+  }
+
+  const body = await response.json();
+  // TODO: Do something with the response.
+
+  rl.close();
+});
+```
+
+There is a bug here. `response.body` is a `Promise` not a string. So if a user ever tries to fetch a URL which returns a non-200 status code, our program will crash:
+
+```console
+% node fetch.js
+What URL should we fetch?
+> http://www.google.com/beepboop
+file:///Users/dwh/tmp/jsplay/fetch.js:12
+    if (response.body.toLowerCase().includes("permission")) {
+                      ^
+
+TypeError: response.body.toLowerCase is not a function
+    at file:///Users/dwh/tmp/jsplay/fetch.js:12:23
+    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
+
+Node.js v22.11.0
+```
+
+The code we wrote was wrong. It could never have been correct. After a `fetch`, `response.body.toLowerCase()` _never_ makes sense. Ideally we shouldn't have needed to wait until running the code (in fact, running exactly that line of code with exactly that data) to find this out.
+
+## Types
+
+This is where types come in.
+
+Imagine if we could run some analysis over our code that told us "You're calling `double` with a string, but `double` expects a number, you have a bug". Or told us "You're calling `response.body.toLowerCase()` but `response.body` is a `Promise` which doesn't have a method `toLowerCase`, you have a bug".
+
+Now we wouldn't need to keep running our program with lots of different inputs every time we change it. The type analysis could tell us "You have a bug here, you should fix it". Without having to run the program, and without having to think about different possible inputs.
+
+## Support for type checking
+
+Different languages have different levels of support for checking types.
+
+Some languages, like Java, C++, Rust, and Go, _require_ you to write what types you expect function parameters to have.
+
+Other languages, like JavaScript and Python, don't _require_ this but they have tools which _allow_ you to add this information. They perform good checks when you do add this type information. But allow you not to add this information to all of your code, and will often ignore code that doesn't have type annotations (which means they may not find all of your type errors).
```

Comment:
```suggestion
Other languages, like JavaScript and Python, don't _require_ this but they have tools which _allow_ you to add this information, like mypy, Typescript, and JSDoc. JS and Python perform good checks when you do add "typing". But in JS and Python you can choose to add types to only parts of your code. The languages will infer the rest of the types as usual. Allowing a mix of typed and untyped code means they may not find all of your type errors.
```

Hmmm, I'm not happy with that either tbh.

Reply:
I did a different rewording, WDYT?